tests/components/ipp/test_config_flow.py

```diff
@@ -0,0 +1,306 @@
+from homeassistant.components.ipp.const import CONF_BASE_PATH, CONF_UUID
+from homeassistant.config_entries import SOURCE_USER, SOURCE_ZEROCONF
+from homeassistant.const import CONF_HOST, CONF_NAME, CONF_SSL
+from homeassistant.core import HomeAssistant
+
+from . import (
+    MOCK_USER_INPUT,
+    MOCK_ZEROCONF_IPP_SERVICE_INFO,
+    MOCK_ZEROCONF_IPPS_SERVICE_INFO,
+    init_integration,
+    load_fixture_binary,
+)
+
+from tests.test_util.aiohttp import AiohttpClientMocker
+
+
+async def test_show_user_form(hass: HomeAssistant) -> None:
+    """Test that the user set up form is served."""
+    result = await hass.config_entries.flow.async_init(
+        config_flow.DOMAIN, context={"source": SOURCE_USER},
+    )
+
+    assert result["step_id"] == "user"
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+
+
+async def test_show_zeroconf_confirm_form(hass: HomeAssistant) -> None:
+    """Test that the zeroconf confirmation form is served."""
+    flow = config_flow.IPPFlowHandler()
+    flow.hass = hass
+    flow.context = {"source": SOURCE_ZEROCONF}
+    flow.discovery_info = {CONF_NAME: "EPSON123456"}
+
+    result = await flow.async_step_zeroconf_confirm()
+
+    assert result["step_id"] == "zeroconf_confirm"
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["description_placeholders"] == {CONF_NAME: "EPSON123456"}
+
+
+async def test_show_zeroconf_form(
+    hass: HomeAssistant, aioclient_mock: AiohttpClientMocker
+) -> None:
+    """Test that the zeroconf confirmation form is served."""
+    aioclient_mock.post(
+        "http://EPSON123456.local:631/ipp/print",
+        content=load_fixture_binary("ipp/get-printer-attributes.bin"),
+        headers={"Content-Type": "application/ipp"},
+    )
+
+    flow = config_flow.IPPFlowHandler()
+    flow.hass = hass
+    flow.context = {"source": SOURCE_ZEROCONF}
+
+    discovery_info = MOCK_ZEROCONF_IPP_SERVICE_INFO.copy()
+    result = await flow.async_step_zeroconf(discovery_info)
+
+    assert flow.discovery_info[CONF_HOST] == "EPSON123456.local"
+    assert flow.discovery_info[CONF_NAME] == "EPSON123456"
+
+    assert result["step_id"] == "zeroconf_confirm"
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["description_placeholders"] == {CONF_NAME: "EPSON123456"}
+
+
+async def test_connection_error(
+    hass: HomeAssistant, aioclient_mock: AiohttpClientMocker
+) -> None:
+    """Test we show user form on IPP connection error."""
+    aioclient_mock.post(
+        "http://EPSON123456.local:631/ipp/print", exc=aiohttp.ClientError
+    )
+
+    user_input = MOCK_USER_INPUT.copy()
+    result = await hass.config_entries.flow.async_init(
+        config_flow.DOMAIN, context={"source": SOURCE_USER}, data=user_input,
+    )
+
+    assert result["step_id"] == "user"
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["errors"] == {"base": "connection_error"}
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5274251)